hsm/manager_hsm.go

```diff
@@ -0,0 +1,311 @@
+	//	  contains support for EdDSA.
+
+	default:
+		return nil, errors.WithStack(jwk.ErrUnsupportedKeyAlgorithm)
+	}
+}
+
+func (m *KeyManager) GetKey(_ context.Context, set, kid string) (*jose.JSONWebKeySet, error) {
+	m.RLock()
+	defer m.RUnlock()
+
+	keyPair, err := m.FindKeyPair([]byte(kid), []byte(set))
+	if err != nil {
+		return nil, err
+	}
+
+	if keyPair == nil {
+		return nil, errors.WithStack(x.ErrNotFound)
+	}
+
+	id, alg, use, err := getKeySetAttributes(m, keyPair, []byte(kid))
+	if err != nil {
+		return nil, err
+	}
+
+	return createKeySet(keyPair, id, alg, use)
+}
+
+func (m *KeyManager) GetKeySet(_ context.Context, set string) (*jose.JSONWebKeySet, error) {
+	m.RLock()
+	defer m.RUnlock()
+
+	keyPairs, err := m.FindKeyPairs(nil, []byte(set))
+	if err != nil {
+		return nil, err
+	}
+
+	if keyPairs == nil {
+		return nil, errors.WithStack(x.ErrNotFound)
+	}
+
+	var keys []jose.JSONWebKey
+	for _, keyPair := range keyPairs {
+		kid, alg, use, err := getKeySetAttributes(m, keyPair, nil)
+		if err != nil {
+			return nil, err
+		}
+		keys = append(keys, createKeys(keyPair, kid, alg, use)...)
+	}
+
+	return &jose.JSONWebKeySet{
+		Keys: keys,
+	}, nil
+}
+
+func (m *KeyManager) DeleteKey(_ context.Context, set, kid string) error {
+	m.Lock()
+	defer m.Unlock()
+
+	keyPair, err := m.FindKeyPair([]byte(kid), []byte(set))
+	if err != nil {
+		return err
+	}
+
+	if keyPair != nil {
+		err = keyPair.Delete()
+		if err != nil {
+			return err
+		}
+	} else {
+		return errors.WithStack(x.ErrNotFound)
+	}
+	return nil
+}
+
+func (m *KeyManager) DeleteKeySet(_ context.Context, set string) error {
+	m.Lock()
+	defer m.Unlock()
+
+	keyPairs, err := m.FindKeyPairs(nil, []byte(set))
+	if err != nil {
+		return err
+	}
+
+	if keyPairs == nil {
+		return errors.WithStack(x.ErrNotFound)
+	}
+
+	for _, keyPair := range keyPairs {
+		err = keyPair.Delete()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *KeyManager) AddKey(_ context.Context, _ string, _ *jose.JSONWebKey) error {
+	return errors.WithStack(ErrPreGeneratedKeys)
+}
+
+func (m *KeyManager) AddKeySet(_ context.Context, _ string, _ *jose.JSONWebKeySet) error {
+	return errors.WithStack(ErrPreGeneratedKeys)
+}
+
+func (m *KeyManager) UpdateKey(_ context.Context, _ string, _ *jose.JSONWebKey) error {
+	return errors.WithStack(ErrPreGeneratedKeys)
+}
+
+func (m *KeyManager) UpdateKeySet(_ context.Context, _ string, _ *jose.JSONWebKeySet) error {
+	return errors.WithStack(ErrPreGeneratedKeys)
+}
+
+func getKeySetAttributes(m *KeyManager, key crypto11.Signer, kid []byte) (string, string, string, error) {
+	if kid == nil {
+		ckaId, err := m.GetAttribute(key, crypto11.CkaId)
+		if err != nil {
+			return "", "", "", err
+		}
+		kid = ckaId.Value
+	}
+
+	var alg string
+	switch k := key.Public().(type) {
+	case *rsa.PublicKey:
+		alg = "RS256"
+		// TODO: Should we validate minimal key length by checking CKA_MODULUS_BITS?
```

Comment:
@aarmam do you want to address this one? Otherwise we can remove the TODO item I think

Reply:
When HSM mode is enabled and no keys are found then RSA keys are generated with key length 4096. But if keys are generated on HSM beforehand then there is no key limit set. We could throw error here if key length is too small. I'm on vacation right now and not able to write serious code :) So I would remove this TODO right now if you think its not important or leave this in and I will do separate PR for this.

Reply:
Ok, that sounds good to me! I will create a follow up issue for it and link it here :)